avail/src/main/kotlin/avail/interpreter/execution/AvailLoader.kt

```diff
@@ -882,17 +883,50 @@ constructor(
 	 * corresponding declarations occur.
 	 */
 	@GuardedBy("styledRangesLock")
-	val usesToDefinitions = RunTree<LongRange>()
+	val usesToDeclarations = RunTree<LongRange>()
 
 	/**
-	 * Access the [usesToDefinitions] in the [action] while holding the lock.
+	 * A [MutableSet] of every file range containing the primary token of a
+	 * local [declaration][DeclarationPhraseDescriptor].
+	 */
+	@GuardedBy("styledRangesLock")
+	val allDeclarations = mutableSetOf<LongRange>()
+
+	/**
+	 * Access the [usesToDeclarations] in the [action] while holding the lock.
+	 *
+	 * @param action
+	 *   The action to perform with [usesToDeclarations] while holding the lock.
+	 */
+	@ThreadSafe
+	fun <T> lockUsesToDeclarations(action: RunTree<LongRange>.()->T): T =
+		styledRangesLock.safeWrite { usesToDeclarations.action() }
+
+	/**
+	 * Access the [allDeclarations] in the [action] while holding the lock.
 	 *
 	 * @param action
-	 *   The action to perform with [usesToDefinitions] while holding the lock.
+	 *   The action to perform with [allDeclarations] while holding the lock.
 	 */
 	@ThreadSafe
-	fun <T> lockUsesToDefinitions(action: RunTree<LongRange>.()->T): T =
-		styledRangesLock.safeWrite { usesToDefinitions.action() }
+	fun <T> lockAllDeclarations(action: MutableSet<LongRange>.()->T): T =
+		styledRangesLock.safeWrite { allDeclarations.action() }
+
+	/**
+	 * A [declaration][DeclarationPhraseDescriptor] phrase was encountered.
+	 * Record its range in the declarationToken [RunTree].  This construct
+	 * exists solely to detect unused declarations.  The range runs from the
+	 * one-based start through the one-based end.
+	 */
+	fun addDeclarationToken(declarationToken: A_Token)
```

Comment:
Anywhere I can look up naming conventions such as `A_Token`, indentation, etc., we follow for the project? thanks!

Reply:
For historic context, all Avail objects have always been of exactly type `AvailObject` ("type" in the Kotlin sense, or the Java or C++ or Smalltalk senses before that).  The host language's own type system has always been far too simplistic to model Avail types directly, which is why `AvailObject` delegates its behavior to its `Descriptor`, passing the `AvailObject` as the first argument of the `AbstractDescriptor.o_*()` operations that encode the real behavior (we sometimes refer to these as subprimitives).
At one point while the VM was in Java, mumble-mumble years ago, I came up with the idea to add the "`A_`*" *superinterfaces* to AvailObject.  If you know some variable or return value is a set, you can weaken its declaration to `A_Set`, which prevents you from accidentally using the `A_Tuple.tupleAt(Int)` subscripting operation on it – which would throw an `AvailUnsupportedOperationException` exception and crash the runtime.   There's nothing forcing you to use the `A_`* types, but it sure helps with things like autosuggestion and minimizing those kind of crashes.
At the top is `A_BasicObject`, which has relatively little behavior defined for it.  Below that are the `A_*` interfaces like `A_Set`, `A_Tuple`, `A_Atom`, `A_RawFunction`, `A_Function`, `A_Phrase`, and I believe a couple dozen more.  These are almost all flat, with the most prominent exception being `A_Tuple` having `A_String` below it, since an Avail string is the same thing as an Avail tuple of characters.
Below the `A_*` interfaces is the abstract class `AvailObjectRepresentation`, which provides basic things like slot access.  Below that is the only concrete class in that hierarchy, `AvailObject`.  If you're given some object and you don't expect to do much with it beyond storing it in another data structure or checking equality or hashing, you should use `A_BasicObject`, the most general interface.  If you're extracting an element of something like a tuple or set, use `AvailObject`, leaving it up to the client to weaken the type appropriately if they know they're working with, say, a tuple of functions, or a set of types.  Occasionally you'll need to cast an `A_BasicObject` into a stronger `A_*`, or even `AvailObject`, but those occurrences are pretty rare (and *definitely* not a performance problem).

For indentation within Avail, we use leading tabs exclusively, other than a "`\t\t * `" in multi-line `/* */`  comments.  We try to conform to the existing code base's style as closely as possible, but I don't believe we have a formal statement about what it is.  We do have plans to introduce indentation rules for methods, keeping them as modular as possible, akin to the modular user-defined styling mechanism.  Not every DSL *should* be styled the same way, but within a particular DSL, it's important to style things consistently.  It was only a couple of years ago that I found a misleadingly-indented else clause in Kotlin code that was the actual source of a bug.  That should be forbidden in most DSLs.  Also, automatic formatting after code transformations would be a nice thing to have.  And although Todd probably raises good points against me on this point, I think it would be nice to automatically wrap expressions to the width of the text window containing them, using the wrapping and indentation rules, preferences, and codified heuristics of the DSL.

For indentation in Kotlin, we also try to conform to the current code base as much as possible, although IntelliJ applies conflicting styles when pasting code, refactoring to inline an expression, or even closing a `}`.  It has multiple formatters, each with their own mechanisms and triggers, and they disagree violently at times.  Perhaps we'll be able to reuse some of the general, modular formatter that we plan to build and apply it to Kotlin as well.  We've had many discussions over the years about how best to recreate Kotlin as an Avail DSL, whether for some automatic code generation, extension of Kotlin to include a macro engine, or to have a mechanical translation of the Avail VM to Rust.  The translation from Smalltalk to Java was mostly mechanical, and in theory we could have kept both VMs runnable for years.

Reply:
Thank you for the detailed answer, @markATavail. This helps me better understand how Avail is modeled and the hierarchy used in Kotlin. Regarding code style, I agree, it would be nice to have a formatter or some way to automate the enforcement of these rules.